File: litho-core/src/main/java/com/facebook/litho/Column.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho;

import com.facebook.litho.annotations.Prop;
import com.facebook.yoga.YogaAlign;
import com.facebook.yoga.YogaFlexDirection;
import com.facebook.yoga.YogaJustify;
import com.facebook.yoga.YogaWrap;
import java.util.ArrayList;
import java.util.List;
import javax.annotation.Nullable;

/** A {@link Component} that renders its children in a column. */
public final class Column extends Component {

  @Nullable
  @Prop(optional = true)
  private List<Component> children;

  @Nullable
  @Prop(optional = true)
  private YogaAlign alignItems;

  @Nullable
  @Prop(optional = true)
  private YogaAlign alignContent;

  @Nullable
  @Prop(optional = true)
  private YogaJustify justifyContent;

  @Nullable
  @Prop(optional = true)
  private YogaWrap wrap;

  @Prop(optional = true)
  private boolean reverse;

  Column(String simpleName) {
    super(simpleName);
  }

  @Override
  protected boolean canResolve() {
    return true;
  }

  public static Builder create(ComponentContext context) {
    return create(context, 0, 0, "Column");
  }

  public static Builder create(ComponentContext context, String simpleName) {
    return create(context, 0, 0, simpleName);
  }

  public static Builder create(ComponentContext context, int defStyleAttr, int defStyleRes) {
    return create(context, defStyleAttr, defStyleRes, "Column");
  }

  public static Builder create(
      ComponentContext context, int defStyleAttr, int defStyleRes, String simpleName) {
    final Builder builder = new Builder();
    builder.init(context, defStyleAttr, defStyleRes, new Column(simpleName));
    return builder;
  }

  @Override
  protected Component onCreateLayout(ComponentContext c) {
    return this;
  }

  @Override
  protected ComponentLayout resolve(ComponentContext c) {
    InternalNode node =
        InternalNodeUtils.create(c)
            .flexDirection(reverse ? YogaFlexDirection.COLUMN_REVERSE : YogaFlexDirection.COLUMN);

    if (alignItems != null) {
      node.alignItems(alignItems);
    }

    if (alignContent != null) {
      node.alignContent(alignContent);
    }

    if (justifyContent != null) {
      node.justifyContent(justifyContent);
    }

    if (wrap != null) {
      node.wrap(wrap);
    }

    if (children != null) {
      for (Component child : children) {
        if (c.wasLayoutCanceled()) {
          return ComponentContext.NULL_LAYOUT;
        }

        if (c.wasLayoutInterrupted()) {
          node.appendUnresolvedComponent(child);
        } else {
          node.child(child);
        }
      }
    }

    return node;
  }

  @Override
  public boolean isEquivalentTo(Component other, boolean shouldCompareState) {
    if (this == other) {
      return true;
    }
    if (other == null || getClass() != other.getClass()) {
      return false;
    }
    Column column = (Column) other;
    if (this.getId() == column.getId()) {
      return true;
    }
    if (children != null) {
      if (column.children == null || children.size() != column.children.size()) {
        return false;
      }
      for (int i = 0, size = children.size(); i < size; i++) {
        if (!children.get(i).isEquivalentTo(column.children.get(i), shouldCompareState)) {
          return false;
        }
      }
    } else if (column.children != null) {
      return false;
    }
    if (alignItems != null ? !alignItems.equals(column.alignItems) : column.alignItems != null) {
      return false;
    }
    if (alignContent != null
        ? !alignContent.equals(column.alignContent)
        : column.alignContent != null) {
      return false;
    }
    if (justifyContent != null
        ? !justifyContent.equals(column.justifyContent)
        : column.justifyContent != null) {
      return false;
    }
    if (reverse != column.reverse) {
      return false;
    }
    return true;
  }

  public static class Builder extends Component.ContainerBuilder<Builder> {
    Column mColumn;
    ComponentContext mContext;

    void init(ComponentContext context, int defStyleAttr, int defStyleRes, Column column) {
      super.init(context, defStyleAttr, defStyleRes, column);
      mColumn = column;
      mContext = context;
    }

    @Override
    protected void setComponent(Component component) {
      mColumn = (Column) component;
    }

    @Override
    public Builder child(@Nullable Component child) {
      if (child == null) {
        return this;
      }

      if (this.mColumn.children == null) {
        this.mColumn.children = new ArrayList<>();
      }

      this.mColumn.children.add(child);
      return this;
    }

    @Override
    public Builder child(@Nullable Component.Builder<?> child) {
      if (child == null) {
        return this;
      }
      return child(child.build());
    }

    @Override
    public Builder alignContent(YogaAlign alignContent) {
      this.mColumn.alignContent = alignContent;
      return this;
    }

    @Override
    public Builder alignItems(YogaAlign alignItems) {
      this.mColumn.alignItems = alignItems;
      return this;
    }

    @Override
    public Builder justifyContent(YogaJustify justifyContent) {
      this.mColumn.justifyContent = justifyContent;
      return this;
    }

    @Override
    public Builder wrap(YogaWrap wrap) {
      this.mColumn.wrap = wrap;
      return this;
    }

    @Override
    public Builder reverse(boolean reverse) {
      this.mColumn.reverse = reverse;
      return this;
    }

    @Override
    public Builder getThis() {
      return this;
    }

    @Override
    public Column build() {
      return mColumn;
    }
  }
}

File: litho-core/src/main/java/com/facebook/litho/Column.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho

import com.facebook.litho.annotations.Prop
import com.facebook.yoga.YogaAlign
import com.facebook.yoga.YogaFlexDirection
import com.facebook.yoga.YogaJustify
import com.facebook.yoga.YogaWrap

/** A [Component] that renders its children in a column. */
class Column
private constructor(
    customSimpleName: String?,
    @field:Prop(optional = true) private var alignContent: YogaAlign?,
    @field:Prop(optional = true) private var alignItems: YogaAlign?,
    @field:Prop(optional = true) private var justifyContent: YogaJustify?,
    @field:Prop(optional = true) private var wrap: YogaWrap?,
    @field:Prop(optional = true) private var reverse: Boolean,
    @field:Prop(optional = true) private var children: MutableList<Component>?,
) : SpecGeneratedComponent(customSimpleName ?: "Column") {

  constructor(
      customSimpleName: String?
  ) : this(customSimpleName, null, null, null, null, false, null)

  @JvmOverloads
  constructor(
      alignContent: YogaAlign?,
      alignItems: YogaAlign?,
      justifyContent: YogaJustify?,
      wrap: YogaWrap?,
      reverse: Boolean,
      children: MutableList<Component>? = null
  ) : this(null, alignContent, alignItems, justifyContent, wrap, reverse, children)

  override fun canResolve(): Boolean = true

  public override fun resolve(resolveContext: ResolveContext, c: ComponentContext): LithoNode? {
    val node = LithoNode()
    node.flexDirection(if (reverse) YogaFlexDirection.COLUMN_REVERSE else YogaFlexDirection.COLUMN)
    alignItems?.let { node.alignItems(it) }
    alignContent?.let { node.alignContent(it) }
    justifyContent?.let { node.justifyContent(it) }
    wrap?.let { node.wrap(it) }
    children?.let { children ->
      for (child in children) {
        if (resolveContext.isFutureReleased) {
          return null
        }
        if (resolveContext.isLayoutInterrupted) {
          node.appendUnresolvedComponent(child)
        } else {
          node.child(resolveContext, c, child)
        }
      }
    }
    return node
  }

  public override fun resolve(
      resolveContext: ResolveContext,
      scopedComponentInfo: ScopedComponentInfo,
      parentWidthSpec: Int,
      parentHeightSpec: Int,
      componentsLogger: ComponentsLogger?
  ): ComponentResolveResult {
    val lithoNode = resolve(resolveContext, scopedComponentInfo.context)
    return ComponentResolveResult(lithoNode, commonProps)
  }

  override fun isEquivalentProps(other: Component?, shouldCompareCommonProps: Boolean): Boolean {
    if (this === other) {
      return true
    }
    if (other !is Column) {
      return false
    }
    if (this.id == other.id) {
      return true
    }
    val children = children
    val otherChildren = other.children
    if (children != null) {
      if (otherChildren == null || children.size != otherChildren.size) {
        return false
      }
      children.forEachIndexed { index, child ->
        if (!child.isEquivalentTo(otherChildren[index], shouldCompareCommonProps)) {
          return false
        }
      }
    } else if (otherChildren != null) {
      return false
    }
    if (alignItems != other.alignItems) {
      return false
    }
    if (alignContent != other.alignContent) {
      return false
    }
    if (justifyContent != other.justifyContent) {
      return false
    }
    return reverse == other.reverse
  }

  override fun usesLocalStateContainer(): Boolean = true

  class Builder
  internal constructor(
      context: ComponentContext,
      defStyleAttr: Int,
      defStyleRes: Int,
      @JvmField var column: Column
  ) : ContainerBuilder<Builder>(context, defStyleAttr, defStyleRes, column) {
    override fun setComponent(component: Component) {
      column = component as Column
    }

    override fun child(child: Component?): Builder {
      if (child == null) {
        return this
      }
      val children = column.children ?: ArrayList<Component>().also { column.children = it }
      children.add(child)
      return this
    }

    override fun child(child: Component.Builder<*>?): Builder =
        if (child == null) {
          this
        } else {
          child(child.build())
        }

    override fun alignContent(alignContent: YogaAlign?): Builder = apply {
      column.alignContent = alignContent
    }

    override fun alignItems(alignItems: YogaAlign?): Builder = apply {
      column.alignItems = alignItems
    }

    override fun justifyContent(justifyContent: YogaJustify?): Builder = apply {
      column.justifyContent = justifyContent
    }

    override fun wrap(wrap: YogaWrap?): Builder = apply { column.wrap = wrap }

    override fun reverse(reverse: Boolean): Builder = apply { column.reverse = reverse }

    override fun getThis(): Builder = this

    override fun build(): Column = column
  }

  companion object {
    @JvmStatic
    fun create(context: ComponentContext?, simpleName: String?): Builder =
        create(context, 0, 0, simpleName)

    @JvmStatic
    @JvmOverloads
    fun create(
        context: ComponentContext?,
        defStyleAttr: Int = 0,
        defStyleRes: Int = 0,
        simpleName: String? = null
    ): Builder = Builder(requireNotNull(context), defStyleAttr, defStyleRes, Column(simpleName))
  }
}
